File: Persistence/src/main/java/org/example/CursaDBRepository.java
package org.example;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.model.Cursa;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class CursaDBRepository implements CursaRepositoryInterface {
    private static final Logger logger = LogManager.getLogger(CursaDBRepository.class);
    private final JdbcUtils dbUtils;

    public CursaDBRepository(Properties properties) {
        dbUtils = new JdbcUtils(properties);
        logger.info("CursaDBRepository initialized");
    }

    @Override
    public Long insert(Cursa entity) {
        logger.info("Inserting new Cursa");
        Connection conn = dbUtils.getConnection();

        try (PreparedStatement prepStmt = conn.prepareStatement(
                "INSERT INTO Cursa (numarParticipanti, capMotor) VALUES (?, ?)",
                Statement.RETURN_GENERATED_KEYS)) {
            prepStmt.setInt(1, entity.getNumarParticipanti());
            prepStmt.setInt(2, entity.getCapMotor());
            prepStmt.executeUpdate();

            try (ResultSet generatedKeys = prepStmt.getGeneratedKeys()) {
                if (generatedKeys.next()) {
                    long id = generatedKeys.getLong(1);
                    logger.info("Inserted new Cursa with id {}", id);
                    return id;
                }
            }
        } catch (SQLException e) {
            logger.error("Error inserting Cursa: " + e.getMessage());
        }
        return -1L;
    }

    @Override
    public void updateById(Long id, Cursa entity) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void deleteById(Long id) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Cursa getById(Long id) {
        logger.info("Fetching Cursa with id {}", id);
        Connection conn = dbUtils.getConnection();

        try (PreparedStatement prepStmt = conn.prepareStatement("SELECT * FROM cursa WHERE id = ?")) {
            prepStmt.setLong(1, id);
            try (ResultSet rs = prepStmt.executeQuery()) {
                if (rs.next()) {
                    return new Cursa(rs.getLong("id"), rs.getInt("numarParticipanti"), rs.getInt("capMotor"));
                }
            }
        } catch (SQLException e) {
            logger.error("Error fetching Cursa: " + e.getMessage());
        }
        return null;
    }

    @Override
    public List<Cursa> getAll() {
        logger.info("Fetching all Curse");
        List<Cursa> curse = new ArrayList<>();
        Connection conn = dbUtils.getConnection();

        try (PreparedStatement prepStmt = conn.prepareStatement("SELECT * FROM cursa");
             ResultSet rs = prepStmt.executeQuery()) {

            while (rs.next()) {
                curse.add(new Cursa(rs.getLong("id"), rs.getInt("numarParticipanti"), rs.getInt("capMotor")));
            }
        } catch (SQLException e) {
            logger.error("Error fetching all Curse: " + e.getMessage());
        }
        logger.info("Returning " + curse.size() + " curse");
        return curse;
    }

    @Override
    public List<Integer> getCapacitati() {
        logger.info("Fetching all motor capacities");
        List<Integer> capacitati = new ArrayList<>();
        Connection conn = dbUtils.getConnection();

        try (PreparedStatement prepStmt = conn.prepareStatement("SELECT DISTINCT capMotor FROM cursa");
             ResultSet rs = prepStmt.executeQuery()) {

            while (rs.next()) {
                capacitati.add(rs.getInt("capMotor"));
            }
        } catch (SQLException e) {
            logger.error("Error fetching capacities: " + e.getMessage());
            throw new RuntimeException("Error fetching capacities");
        }
        return capacitati;
    }

    @Override
    public void increaseNr(Long id) {
        logger.info("Increasing numarParticipanti for Cursa with id {}", id);
        Connection conn = dbUtils.getConnection();

        try (PreparedStatement prepStmt = conn.prepareStatement(
                "UPDATE cursa SET numarParticipanti = numarParticipanti + 1 WHERE id = ?")) {
            prepStmt.setLong(1, id);
            prepStmt.executeUpdate();
        } catch (SQLException e) {
            logger.error("Error increasing numarParticipanti for id {}: {}", id, e.getMessage());
        }
    }

    @Override
    public List<Cursa> getCurseByCap(int cap) {
        logger.info("Fetching all Curse with motor capacity {}", cap);
        List<Cursa> curse = new ArrayList<>();
        Connection conn = dbUtils.getConnection();

        try (PreparedStatement prepStmt = conn.prepareStatement("SELECT * FROM cursa WHERE capMotor = ?")) {
            prepStmt.setInt(1, cap);
            try (ResultSet rs = prepStmt.executeQuery()) {
                while (rs.next()) {
                    curse.add(new Cursa(rs.getLong("id"), rs.getInt("numarParticipanti"), rs.getInt("capMotor")));
                }
            }
        } catch (SQLException e) {
            logger.error("Error fetching Curse by capMotor: " + e.getMessage());
        }
        logger.info("Returning " + curse.size() + " curse");
        return curse;
    }
}

File: persistence/CursaDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Data.Sqlite;
using log4net;
using model;

namespace persistence
{
    public class CursaDBRepository : IRepositoryCursa
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(CursaDBRepository));
        private readonly DBUtils _dbUtils;

        public CursaDBRepository(IDictionary<string, string> properties)
        {
            _dbUtils = new DBUtils(properties);
            logger.Info("RepositoryDBUser initialized");
        }

        public List<int> getCapacitati()
        {
            logger.Info("fetching all team names");
            List<int> capacitati = new List<int>();

            using (IDbConnection connection = _dbUtils.getConnection())
            {
                try
                {
                    if (connection is SqliteConnection sqliteConn)
                    {
                        string query = "SELECT DISTINCT capMotor FROM Cursa";

                        using (var cmd = new SqliteCommand(query, sqliteConn))
                        {
                            using (var reader = cmd.ExecuteReader())
                            {
                                while (reader.Read())
                                    capacitati.Add(reader.GetInt32(0));
                            }
                        }
                        return capacitati;
                    }
                }
                catch (Exception e)
                {
                    logger.Error("eroare la getEchipe " + e.Message);
                    throw new Exception("eroare la listarea echipelor");
                }
            }
            return capacitati;
        }

        public long Insert(Cursa entity)
        {
            logger.Info("Inserting new Cursa");

            using (IDbConnection conn = _dbUtils.getConnection())
            {
                try
                {
                    if (conn is SqliteConnection sqliteConn)
                    {
                        string query = "INSERT INTO Cursa (numarParticipanti, capMotor) VALUES (@numarParticipanti, @capMotor); SELECT last_insert_rowid();";

                        using (var cmd = new SqliteCommand(query, sqliteConn))
                        {
                            cmd.Parameters.AddWithValue("@numarParticipanti", entity.NumarParticipanti);
                            cmd.Parameters.AddWithValue("@capMotor", entity.CapMotor);

                            long id = (long)cmd.ExecuteScalar();
                            logger.Info($"Inserted new Cursa with id {id}");
                            return id;
                        }
                    }
                }
                catch (Exception ex)
                {
                    logger.Error("Error inserting Cursa: " + ex.Message);
                }
            }
            return -1;
        }


        public void UpdateById(long id, Cursa entity)
        {
            throw new NotImplementedException();
        }

        public void DeleteById(long id)
        {
            throw new NotImplementedException();
        }

        public Cursa GetById(long id)
        {
            logger.Info($"Fetching Cursa with id {id}");

            using (IDbConnection conn = _dbUtils.getConnection())
            {
                try
                {
                    if (conn is SqliteConnection sqliteConn)
                    {
                        string query = "SELECT * FROM Cursa WHERE id = @id";

                        using (var cmd = new SqliteCommand(query, sqliteConn))
                        {
                            cmd.Parameters.AddWithValue("@id", id);
                            using (var reader = cmd.ExecuteReader())
                            {
                                if (reader.Read())
                                {
                                    return new Cursa(reader.GetInt64(0), reader.GetInt32(1), reader.GetInt32(2));
                                }
                            }
                        }
                    }
                }
                catch (Exception ex)
                {
                    logger.Error("Error fetching Cursa: " + ex.Message);
                }
            }
            return null;
        }

        public void increaseNr(long id)
        {
            logger.Info("incerasing nr of participanti for " + id);

            using (IDbConnection conn = _dbUtils.getConnection())
            {
                try
                {
                    if (conn is SqliteConnection sqliteConn)
                    {
                        string query = "UPDATE Cursa SET numarParticipanti = numarParticipanti + 1 WHERE id = @id";

                        using (var cmd = new SqliteCommand(query, sqliteConn))
                        {
                            cmd.Parameters.AddWithValue("@id", id);
                            cmd.ExecuteNonQuery();
                        }
                    }
                }
                catch (Exception ex)
                {
                    logger.Error("Error incerasing nr of participanti for " + id + " : " + ex.Message);
                }
            }
        }

        public List<Cursa> GetAll()
        {
            logger.Info("Fetching all Curse");

            List<Cursa> curse = new List<Cursa>();

            using (IDbConnection conn = _dbUtils.getConnection())
            {
                try
                {
                    if (conn is SqliteConnection sqliteConn)
                    {
                        string query = "SELECT * FROM Cursa";

                        using (var cmd = new SqliteCommand(query, sqliteConn))
                        {
                            using (var reader = cmd.ExecuteReader())
                            {
                                while (reader.Read())
                                {
                                    curse.Add(new Cursa(reader.GetInt64(0), reader.GetInt32(1), reader.GetInt32(2)));
                                }
                            }
                        }
                    }
                }
                catch (Exception ex)
                {
                    logger.Error("Error fetching all Curse: " + ex.Message);
                }
            }
            return curse;
        }

        public List<Cursa> getCurseByCap(int cap)
        {
            logger.Info("Fetching all Curse with motor capacity " + cap);

            List<Cursa> curse = new List<Cursa>();

            using (IDbConnection conn = _dbUtils.getConnection())
            {
                try
                {
                    if (conn is SqliteConnection sqliteConn)
                    {
                        string query = "SELECT * FROM Cursa WHERE capMotor = @cap";

                        using (var cmd = new SqliteCommand(query, sqliteConn))
                        {
                            cmd.Parameters.AddWithValue("@cap", cap);
                            using (var reader = cmd.ExecuteReader())
                            {
                                while (reader.Read())
                                {
                                    curse.Add(new Cursa(reader.GetInt64(0), reader.GetInt32(1), reader.GetInt32(2)));
                                }
                            }
                        }
                    }
                }
                catch (Exception ex)
                {
                    logger.Error("Error fetching all Curse: " + ex.Message);
                }
            }
            return curse;
        }

    }

}
